EASY-2003 validate example XML's with schema's

jo-pol opened the pull request:
#### When applied it will
* validate examples with the last local schema's
* validate examples with public schema's but ignore these tests while the last version is under development and not yet published
* check that unqualified versions of schema's equal the last qualified version
* [x] add optional attribute 'user-account' to test `dcx-dai/example2.xml `
* [x] code formatting

#### Where should the reviewer @DANS-KNAW/easy start?

#### How should this be manually tested?

Create `target/easy-schema`:
* synchronise https://github.com/DANS-KNAW/easy-schema locally and run `mvn clean install`
* run `mvn clean install -DskipTests=true`

_Should this procedure be added to the README?_

#### Related pull requests on github
repo                       | PR                | note
-------------------------- | ----------------- | ----
easy-                      | [PR#](PRlink)     | ..

Commit:
WIP

jo-pol committed:
fix/ignore tests to run with easy-schema pull request 68

jo-pol committed:
centralized println of error context

jo-pol committed:
a test class per main schema

jo-pol committed:
test fixed with easy-scheme pull request 67 (merged after 68)

jo-pol committed:
easy schemas loaded with local locations; refactoring

jo-pol committed:
validation with TableDriven only; only top-level schema-locations

jo-pol committed:
table driven test: last versioned equals version-less

jo-pol committed:
update agreement example

jo-pol committed:
last dai xsd, not hard coded

jo-pol committed:
renames

jo-pol committed:
validation with not qualified public schema's (when published)

jo-pol committed:
template for pull requests

jo-pol committed:
typo's

jo-pol commented on src/test/scala/nl/knaw/dans/easy/schemaExamples/DefaultsSpec.scala:
```diff
@@ -0,0 +1,51 @@
+      ("md/emd", "qdc.xsd"),
+      ("md/emd", "sdc.xsd"),
+      ("md/emd", "xml.xsd"),
+      ("bag/metadata/files", "files.xsd"),
```
TODO: 
("bag/metadata/agreements", "agreements.xsd")
("bag/metadata/amd", "amd.xsd")
("bag/metadata/amd", "wfs.xsd")

I suppose we don't have examples for `files` and `amd` as we generate them?

_overlooked separate test and amd only has qualified xsd's_

lindareijnhoudt commented:
for files.xml there are examples in the sword2-examples repo. use for instance the file-accessibility example

jo-pol commented:
easy-sword2-examples is missing on http://maven-repo.dans.knaw.nl/repo/nl/knaw/dans/easy/
thus the plugin refuses to add it to `target` like easy-schema.

rvanheest commented:
I would imagine you copy/modify those examples to this project, right? That's what the repository is for. There's no need for importing it to `target`.

jo-pol committed:
example for files.xsd (accessibilities of easy-sword2-dans-examples)

https://github.com/DANS-KNAW/easy-sword2-dans-examples/blob/ddf19e7/src/main/resources/examples/file-accessibilities/metadata/files.xml

jo-pol committed:
renames: tests are organized per schema, not per dir with test files

rvanheest commented on src/main/assembly/dist/examples/abr-type/example1.xml:
```diff
@@ -21,9 +21,7 @@
          xmlns:abr="http://www.den.nl/standaard/166/Archeologisch-Basisregister/"
-         xsi:schemaLocation=
-                 "http://easy.dans.knaw.nl/schemas/md/ddm/ https://easy.dans.knaw.nl/schemas/md/ddm/ddm.xsd
-    http://www.den.nl/standaard/166/Archeologisch-Basisregister/ https://easy.dans.knaw.nl/schemas/vocab/abr-type.xsd">
```
is this no longer necessary?

same for the other instances in this whole diff where you removed `schemaLocation` entries

jo-pol replied:
ddm.xsd explicitly [imports](https://github.com/DANS-KNAW/easy-schema/blob/6f1023cc81b5394afbd2ce216a3db8b3f76ab068/src/main/assembly/dist/md/2019/01/ddm.xsd#L76-L77) these locations. Specifying them here too could only cause contradictions IMHO. Sadly I did not [document](https://github.com/DANS-KNAW/easy-schema-examples/commit/e9dd27f4d8ccf265d7584f33f6cf2d8c00b16ab0) what else triggered me.

rvanheest replied:
👍

rvanheest commented on src/main/assembly/dist/examples/bag/files/example1.xml:
```diff
@@ -0,0 +1,62 @@
+        <visibleToRights>NONE</visibleToRights>
+    </file>
+</files>
+
```
```suggestion
```

rvanheest commented on src/test/scala/nl/knaw/dans/easy/schemaExamples/package.scala:
```diff
@@ -0,0 +1,37 @@
+      .walk()
+      .map(_.toString())
+      .filter(_.matches(".*/[0-9/]{5,}+" + file)) // end with: /YYYY/xxx or /YYYY/MM/xxx
+      .maxBy(_.toUpperCase.map(c => digitsToLowerCase(c))) // digits sort after letters
```
```suggestion
      .maxBy(_.toUpperCase.map(digitsToLowerCase)) // digits sort after letters
```

rvanheest commented on src/test/scala/nl/knaw/dans/easy/schemaExamples/SchemaValidationFixture.scala:
```diff
@@ -15,19 +15,124 @@
+      .map(s => s.replace(httpsEasySchemaBase, schemaDir.toString()))
+  }
+
+  private def schemaIsOnline(schema: Try[Schema]) = {
```
```suggestion
  private def schemaIsOnline(schema: Try[Schema]): Boolean = {
```

rvanheest commented on src/test/scala/nl/knaw/dans/easy/schemaExamples/SchemaValidationFixture.scala:
```diff
@@ -15,19 +15,124 @@
+  }
+
+  private def easyLocationsIn(xml: Elem): Seq[String] = {
+    Seq(
```
Not sure, but would a for-comprehension be more clear to read?

```scala
for {
    s <- Seq("xsi:schemaLocation", "xsi:noNamespaceSchemaLocation")
    str <- xml.attributes.asAttrMap.getOrElse(s, "").split(" +")
    if str.endsWith(".xsd") && str.contains("easy.dans")
} yield str.replace(httpsEasySchemaBase, schemaDir.toString())
```

rvanheest commented on src/test/scala/nl/knaw/dans/easy/schemaExamples/SchemaValidationFixture.scala:
```diff
@@ -15,19 +15,124 @@
+    // and the new schema's are not yet published
+    // reported with: ...lastLocalIsPublic was false
+    val newPublicSchema = localSchemaFile.replace(schemaDir.toString(), httpsEasySchemaBase)
+    Try(new URL(newPublicSchema).openStream()) match {
```
`Try(new URL(newPublicSchema).openStream()).fold(_ => false, _ => true)`

rvanheest commented on src/test/scala/nl/knaw/dans/easy/schemaExamples/DefaultsSpec.scala:
```diff
@@ -0,0 +1,68 @@
+  }
+
+  private def countTestedFiles(file: File) = {
+    file.
```
format please 😉

rvanheest reviewed (commented):
this looks great!

jo-pol committed:
review comment

https://github.com/DANS-KNAW/easy-schema-examples/pull/1/files#r265015510

jo-pol committed:
formatting

rvanheest committed:
Apply suggestions from code review

Co-Authored-By: jo-pol <jo-pol@users.noreply.github.com>

rvanheest reviewed (commented):
Code and tests look good.
What will now be the procedure when you add a new feature to `easy-schema` and you want to test (validate) that with the examples in this projects?

jo-pol committed:
refactoring: extraction of XSD locations turned inside out

https://github.com/DANS-KNAW/easy-schema-examples/pull/1#discussion_r265015312
proposed a for-comprehension

jo-pol committed:
documented development cycle

https://github.com/DANS-KNAW/easy-schema-examples/pull/1#pullrequestreview-214370917

rvanheest commented on src/test/scala/nl/knaw/dans/easy/schemaExamples/SchemaValidationFixture.scala:
```diff
@@ -15,19 +15,128 @@
+  private lazy val triedPublicSchema: Try[Schema] = Try {
+    SchemaFactory
+      .newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI)
+      .newSchema(Array(new StreamSource(publicSchema.toString)).toArray[Source])
```
```suggestion
      .newSchema(Array[Source](new StreamSource(publicSchema.toString)))
```

rvanheest commented on src/test/scala/nl/knaw/dans/easy/schemaExamples/SchemaValidationFixture.scala:
```diff
@@ -15,19 +15,128 @@
+    ).inputStream
+    SchemaFactory
+      .newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI)
+      .newSchema(Array(new StreamSource(xsdInputStream)).toArray[Source])
```
```suggestion
      .newSchema(Array[Source](new StreamSource(xsdInputStream)))
```

rvanheest commented on src/test/scala/nl/knaw/dans/easy/schemaExamples/SchemaValidationFixture.scala:
```diff
@@ -15,19 +15,128 @@
+  "examples" should "be schema valid with local copy of easy-schema" in {
+    forEvery(examples) { example =>
+      val xml = XML.loadFile((exampleDir / example).toString())
+      easyLocationsIn(xml).foreach(File(_).toJava should exist)
```
1. if you add `implicit def existenceOfFile[FILE <: better.files.File]: Existence[FILE] = _.exists` to the top of this file (first thing inside the trait), you don't need the `.toJava` anymore
2. would it be nice to have `easyLocationsIn` return `better.files.File`?
3. if you apply the above changes, you can write this line like `all(easyLocationsIn(xml)) should exist`, which would probably give you a better error message if something fails (not fail-fast). Side-note, there is an `all` and an `every`; I'm not sure yet what the difference between them is... feel free to experiment a bit 😉

jo-pol replied:
The all and every popped up before.  All reported only the first broken test, every reports every.

rvanheest replied:
ah, now I remember! Then you should use `every`, right?

jo-pol replied:
My answer applied to line 59, you are talking about 61....

jo-pol replied:
@1 implicit is overkill, it affects the subclasses to.

rvanheest replied:
I don't get what you mean. If you add `implicit def existenceOfFile[FILE <: better.files.File]: Existence[FILE] = _.exists` to the package object and replace `easyLocationsIn(xml).foreach(File(_).toJava should exist)` with `every(easyLocationsIn(xml)) should exist`, you get a way better result, right? Could you clarify what you mean with 

> it affects the subclasses to.

rvanheest commented on src/test/scala/nl/knaw/dans/easy/schemaExamples/SchemaValidationFixture.scala:
```diff
@@ -15,19 +15,128 @@
+  }
+
+  private def validate(schema: Try[Schema], xmlString: String): Try[Unit] = {
+    schema.map(_.newValidator().validate(new StreamSource(xmlString.inputStream))) match {
```
this pattern match sounds like:

```scala
schema.map(_.newValidator().validate(new StreamSource(xmlString.inputStream)))
  .doIfFailure {
    case e: SAXParseException => showErrorWithSourceContext(xmlString, e)
  }
```

it's just a side-effect, you're performing, right?

Only you need to add the following to the `pom.xml` for this:

```xml
<dependency>
    <groupId>nl.knaw.dans.lib</groupId>
    <artifactId>dans-scala-lib_2.12</artifactId>
    <scope>test</scope>
</dependency>
```

rvanheest commented on src/test/scala/nl/knaw/dans/easy/schemaExamples/SchemaValidationFixture.scala:
```diff
@@ -15,19 +15,128 @@
+
+  private def easyLocationsIn(xml: Elem): Seq[String] = {
+
+    val attributeValues = xml.attributes.asAttrMap.map {
```
this sounds like a perfect use case for `collect` 🤔

```scala
val attributeValues = xml.attributes.asAttrMap.collect {
  case ("xsi:noNamespaceSchemaLocation", value: String) => s"dummyUri $value"
  case ("xsi:schemaLocation", value: String) => value
}
```

rvanheest commented on src/test/scala/nl/knaw/dans/easy/schemaExamples/SchemaValidationFixture.scala:
```diff
@@ -15,19 +15,128 @@
+    }.filter(_.isDefined).flatten
+    // now we should have strings formatted like: "nsUri1 xsdUrl1 nsUri2 xsdUrl2"
+    attributeValues
+      .flatMap(_.trim.split(" +").grouped(2).toList)  // extract (uri,url) tuples
```
`.toList` isn't necessary

```suggestion
      .flatMap(_.trim.split(" +").grouped(2))  // extract (uri,url) tuples
```

rvanheest commented on src/test/scala/nl/knaw/dans/easy/schemaExamples/SchemaValidationFixture.scala:
```diff
@@ -15,19 +15,128 @@
+    }
+  }
+
+  private def easyLocationsIn(xml: Elem): Seq[String] = {
```
I've had to spend some time with this function to completely understand what it is doing. May I suggest a little rewrite that hopefully makes it better readable? This includes the 2 comments I made about this function above/below (depending on how GitHub renders things these days 😉)

```scala
private def easyLocationsIn(xml: Elem): Iterable[File] = {
  val attributeValues = xml.attributes.asAttrMap.collect {
    case ("xsi:noNamespaceSchemaLocation", value: String) => s"dummyUri $value"
    case ("xsi:schemaLocation", value: String) => value
  }

  // now we should have strings formatted like: "nsUri1 xsdUrl1 nsUri2 xsdUrl2"
  for {
    value <- attributeValues
    tuples <- value.trim.split(" +").grouped(2) // extract (uri,url) tuples
    url = tuples.applyOrElse(1, "").toString // get second element of each tuple
    if url contains "easy.dans"
  } yield File(url.replace(httpsEasySchemaBase, schemaDir.toString()))
}
```

rvanheest commented on src/test/scala/nl/knaw/dans/easy/schemaExamples/DefaultsSpec.scala:
```diff
@@ -0,0 +1,68 @@
+      .count(!_.isDirectory) shouldBe
+      File("src/test/scala/nl/knaw/dans/easy/schemaExamples")
+        .walk()
+        .filter(_.toString().endsWith("SchemaSpec.scala"))
```
would replacing the `filter` and `map` with `collect` make it more readable here?

rvanheest commented on README.md:
```diff
@@ -5,4 +5,13 @@ easy-schema-examples
 A collection of examples of the XML schema's used by EASY. These schema's are published at 
 [https://easy.dans.knaw.nl/schemas/](https://easy.dans.knaw.nl/schemas/)
 
+Development cycle
```
a couple of thoughts about this section:

1. Great to see such a detailed description! It makes the development cycle very clear.
2. Rather than `mvncist`, use `mvn clean generate-test-resources` (bullet 3).
3. Should there be a mention here of when to set `easy-schema` version numbers in `pom.xml` for `easy.schema.version` to certain values?
4. Regarding the last bullet, saying "_The test ignored above should now succeed._" - is this guaranteed? Are there situations where this test can still fail?

jo-pol replied:
@ 3 Is this a question for @lindareijnhoudt 

@ 4 @lindareijnhoudt requested a test like this.

rvanheest committed:
review suggestions

lindareijnhoudt commented on README.md:
```diff
@@ -5,4 +5,13 @@ easy-schema-examples
+* Run `mvn clean install` in `easy-schema`.
+* Run `mvn clean install -DskipTests=true` in this project.
+* Update the examples and fix failing tests.
+* Some tests will be ignored with `...lastLocalIsPublic was false` messages.
```
is this the `pending until fixed`?

jo-pol replied:
This comment is about [`assume(lastLocalIsPublic)`](https://github.com/DANS-KNAW/easy-schema-examples/blob/457810a7a405827ba95567bf06e2a23ca2aa7615/src/test/scala/nl/knaw/dans/easy/schemaExamples/SchemaValidationFixture.scala#L78-L81) as the message says.

Following test is beyond the scope of this issue. It is about white space differences between a versioned and and last non versioned schema. Don't know how to ignore these differences in a reliable way. Comparing parsed schema's would ignore comment too, as far as I know.
https://github.com/DANS-KNAW/easy-schema-examples/blob/8d73112f3c9966a24e9e6f0056d3fbf4f53334a8/src/test/scala/nl/knaw/dans/easy/schemaExamples/DefaultsSpec.scala#L46-L47

lindareijnhoudt commented on src/main/assembly/dist/examples/dcx-dai/example2.xml:
```diff
@@ -17,7 +17,7 @@
 -->
 <dcx-dai:creatorDetails xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:dcx-dai="http://easy.dans.knaw.nl/schemas/dcx/dai/"
-    xsi:schemaLocation="http://easy.dans.knaw.nl/schemas/dcx/dai/ https://easy.dans.knaw.nl/schemas/dcx/dcx-dai.xsd">
+                        xsi:schemaLocation="http://easy.dans.knaw.nl/schemas/dcx/dai/ https://easy.dans.knaw.nl/schemas/dcx/2019/01/dcx-dai.xsd">
```
do you want a non-versioned schema of this?

jo-pol replied:
It is not a top-level schema, but it is good to test it like this. Perhaps I should inject the last found version while reading like I replace the public http protocol based easy schema's with local file based schema's.

jo-pol committed:
filter.map -> collect (review comment)

https://github.com/DANS-KNAW/easy-schema-examples/pull/1#discussion_r265877870

rvanheest commented on pom.xml:
```diff
@@ -16,7 +16,8 @@
 
 -->
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
use `easy.schema.version = 2.1.4-SNAPSHOT` below

rvanheest commented on README.md:
```diff
@@ -5,4 +5,13 @@ easy-schema-examples
 
+* Add new version of XSD to https://github.com/DANS-KNAW/easy-schema.
+* Run `mvn clean install` in `easy-schema`.
+* Run `mvn clean install generate-test-resources ` in this project.
```
with the change from `mvncist` to `mvn clean install generate-test-resources` you are now in trouble, because it runs the tests, which is not what you want in this stage. Use `mvn clean generate-test-resources` (without the `install`) instead.

```suggestion
* Run `mvn clean generate-test-resources` in this project.
```

rvanheest commented on src/test/scala/nl/knaw/dans/easy/schemaExamples/SchemaValidationFixture.scala:
```diff
@@ -15,19 +15,128 @@
+      }
+  }
+
+  private def easyLocationsIn(xml: Elem): immutable.Iterable[File] = {
```
```suggestion
  private def easyLocationsIn(xml: Elem): Iterable[File] = {
```

rvanheest commented on src/test/scala/nl/knaw/dans/easy/schemaExamples/SchemaValidationFixture.scala:
```diff
@@ -15,19 +15,128 @@
 
-import scala.util.{ Failure, Try }
-import scala.xml.{ Elem, SAXParseException }
+import scala.collection.immutable
```
```suggestion
```

unused import after https://github.com/DANS-KNAW/easy-schema-examples/pull/1/files#r266319875

rvanheest committed:
implicit def existenceOfFile

https://github.com/DANS-KNAW/easy-schema-examples/pull/1#discussion_r266319608

rvanheest committed:
updated schema version

https://github.com/DANS-KNAW/easy-schema-examples/pull/1#discussion_r266316813

rvanheest committed:
Apply suggestions from code review

Co-Authored-By: jo-pol <jo-pol@users.noreply.github.com>

lindareijnhoudt commented on src/main/assembly/dist/examples/bag/agreements/example1.xml:
```diff
@@ -22,7 +22,7 @@
 >
     <depositAgreement>
-        <depositorId>MisterX</depositorId>
+        <signerId>MisterX</signerId>
```
maybe add one with the optional attribute 'user-account'

lindareijnhoudt replied:
```suggestion
        <signerId easy-account="user001">MisterX</signerId>
```

jo-pol committed:
reduce/check maintenance of test code/data (versions)

jo-pol committed:
rewording because of review comment

https://github.com/DANS-KNAW/easy-schema-examples/pull/1#discussion_r266062706

jo-pol committed:
formatting

jo-pol committed:
formatting

lindareijnhoudt committed:
optional attribute 'user-account'

Co-Authored-By: jo-pol <jo-pol@users.noreply.github.com>
